### PR TITLE
Add new sample app component to demonstrate how to add/remove rows.

### PR DESCRIPTION
**Pull request by meisen45**
In this cl, there are several changes:
1. The gitignore file is updated, so non-related file will not be included into this repo.
2. A new package (rows) is added to demonstrate how to allow user to publish/ remove rows from local app.
3. glide package has been updated to 4.0 for better API design and performance.

### CODE REVIEW

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/settings/SettingsExampleFragment.java [hunk 1]
```diff
@@ -30,9 +30,10 @@
 import java.util.Arrays;
 import java.util.Stack;
 
-public class SettingsExampleFragment extends LeanbackSettingsFragment implements DialogPreference.TargetFragment {
+public class SettingsExampleFragment extends LeanbackSettingsFragment
+        implements DialogPreference.TargetFragment {
 
-    private final Stack<Fragment> fragments = new Stack<Fragment>();
+    private static final Stack<Fragment> fragments = new Stack<Fragment>();
```
Comment (dakegu2) [inline app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/settings/SettingsExampleFragment.java]: ?

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/CardPresenter.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import android.widget.PopupMenu;
+
+import com.bumptech.glide.Glide;
+
+/**
+ * A specialized CardPresenter to generate Views and bind Objects to them on demand.
+ *
+ * It will change background color/ card image accordingly
+ */
+public class CardPresenter extends Presenter {
```
Comment (dakegu2) [inline app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/CardPresenter.java]: VideoContentCardPresenter

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+            return channelId;
+        }
+
+        @Override
+        protected Long doInBackground(ChannelContents... params) {
+            return addChannel(PublishChannelFragment.this.getActivity(), params[0]);
+        }
+
+        @Override
+        protected void onPostExecute(Long channelId) {
```
Comment (dakegu2) [inline app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java]: startActivity here means the data layer needs to know which activity handle this, it's better to let data layer call listener in a Handler.

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+         */
+        private String createInputId(Context context) {
+            ComponentName cName = new ComponentName(context, DynamicVideoRowsActivity.class.getName());
+            return TvContractCompat.buildInputId(cName);
+        }
+
+        /**
+         * A transaction id will be assigned to the added channel
+         */
+        private long addChannel(Context context, ChannelContents channelContents) {
```
Comment (dakegu2) [inline app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java]: The code should be inside data layer (ChannelContents), not in the UI code.

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+            } catch (ActivityNotFoundException e) {
+                Log.e(TAG, "could not start add channel approval UI", e);
+            }
+        }
+    }
+
+    /**
+     * Async Task to remove channel from home screen
+     */
+    private final class RemoveChannelInMainScreen extends AsyncTask<ChannelContents, Void, Void> {
```
Comment (dakegu2) [inline app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java]: The code should be inside data layer (ChannelContents), not in the UI code.

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/CardPresenter.java (hunk 2 above)
Comment (meisen45) [reply app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/CardPresenter.java]: Good name, done!

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java (hunk 4 above)
Comment (meisen45) [reply app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java]: yes, done!

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java (hunk 5 above)
Comment (meisen45) [reply app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java]: Yes, done. Will organize the files in a cleaner way next time.

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/settings/SettingsExampleFragment.java (hunk 1 above)
Comment (meisen45) [reply app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/settings/SettingsExampleFragment.java]: Actually our settings example activity has a bug (the app will be crashed once this activity is started). I will fix this bug in an independent CL, some change was put in this CL accidentally, will be more careful next time.

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java (hunk 3 above)
Comment (meisen45) [reply app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java]: It is a great idea, thank you for the detailed explanation. I never considered it before.

Also, for better readability, the LoadAddedChannels and ChannelPlaylistId was extracted as independent file. Please let me know if you think it is not a good idea.

**Review by meisen45 — COMMENTED** [review]
Thank you for your great suggestion, new patch set is uploaded

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Async task to load published channel
+ * So the publish status will be loaded every time when the setting activity is created
+ *
+ * Usage: Just create the object (new LoadAddedChannels()), then the async task will be executed
+ */
+public class LoadAddedChannels extends AsyncTask<Void, Void, Void> {
```
Comment (dakegu2) [inline app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java]: move this class into ChannelContents class

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+     * the context
+     */
+    Context mContext;
+
+    /**
+     * Registered Listener for this aysnc task
+     */
+    private Listener mListener;
+
+    interface Listener {
```
Comment (dakegu2) [inline app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java]: The Listener should be a member of ChannelContents.  The caller who add/remove channels is not necessary the Listener.  For example,  both the BrowseFragment and GuidedStepFragment should add the Listener, whoever causes the data change is irrelevant.

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java (hunk 6 above)
Comment (meisen45) [reply app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java]: Done

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java (hunk 7 above)
Comment (meisen45) [reply app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java]: Yes, you are absolutely right. Thank you very much for the explanation!

**Comment by meisen45** [comment]
A new patch set has been uploaded to address the file organization issue

**Comment by meisen45** [comment]
A new patch set has been uploaded to address the bug we have discussed before

**Comment by meisen45** [comment]
New patch set has been uploaded to address the issue we talked about this morning
Thank you for reviewing my code!

File: app/build.gradle [hunk 8]
```diff
@@ -1,13 +1,13 @@
...
-    compileSdkVersion 'android-O'
-    buildToolsVersion '26.0.0-rc2'
+    compileSdkVersion 'android-26'
+    buildToolsVersion '26.0.0'
 
     defaultConfig {
         applicationId "android.support.v17.leanback.supportleanbackshowcase"
-        minSdkVersion 17
-        targetSdkVersion 'O'
+        minSdkVersion 23
```
Comment (dakegu2) [inline app/build.gradle]: keep minSdkVersion as 17

File: app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/DynamicRowsFragment.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        public void onItemClicked(Presenter.ViewHolder itemViewHolder, Object item,
+                                  RowPresenter.ViewHolder rowViewHolder, Row row) {
+            if (isAdded()) {
+                if (item instanceof VideoContent) {
+                    VideoContent clip = (VideoContent) item;
+                    Intent intent = new Intent(getActivity(), VideoPlaybackActivity.class);
+                    intent.putExtra(VideoExampleActivity.TAG, clip);
+                    startActivity(intent);
+                } else {
+                    Intent intent = new Intent(getActivity(), ChannelPublishActivity.class);
```
Comment (dakegu2) [inline app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/DynamicRowsFragment.java]: show a toast "add to home screen not supported" if sdk version is too low

File: app/build.gradle [hunk 10]
```diff
@@ -22,7 +22,7 @@ android {
 
 dependencies {
     compile fileTree(dir: 'libs', include: ['*.jar'])
-    compile "com.android.support:recyclerview-v7:+"
+    compile "com.android.support:recyclerview-v7:26.0.0-beta2"
```
Comment (dakegu2) [inline app/build.gradle]: no need to change this?

Comment (meisen45) [reply app/build.gradle]: done

File: app/build.gradle (hunk 8 above)
Comment (meisen45) [reply app/build.gradle]: done

**Comment by meisen45** [comment]
New patch set has been uploaded to fix some issue about minsdk version and library version

also a toast will be created if the system version doesn't support "add to home screen" operation

**Review by dakegu2 — APPROVED** [review]

**Pull request merged**